test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,384 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package ipam is a helper to claim ip addresses from an IPAM provider cluster.
+package ipam
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/vim25/mo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	ipamv1 "sigs.k8s.io/cluster-api/exp/ipam/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	. "sigs.k8s.io/cluster-api-provider-vsphere/test/e2e/helper"
+)
+
+var ipamScheme *runtime.Scheme
+
+const controlPlaneEndpointVariable = "CONTROL_PLANE_ENDPOINT_IP"
+
+func init() {
+	ipamScheme = runtime.NewScheme()
+	_ = ipamv1.AddToScheme(ipamScheme)
+}
+
+type IPAddressClaims []*ipamv1.IPAddressClaim
+
+type Helper interface {
+	// ClaimIPs claims IP addresses with the variable name `CONTROL_PLANE_ENDPOINT_IP` and whatever is passed as
+	// additionalIPVariableNames and creates a new clusterctl config file.
+	// It returns the path to the new clusterctl config file and a slice of IPAddressClaims.
+	ClaimIPs(ctx context.Context, clusterctlConfigPath string, additionalIPVariableNames ...string) (localClusterctlConfigFile string, claims IPAddressClaims)
+
+	// Cleanup deletes the given IPAddressClaims.
+	Cleanup(ctx context.Context, claims IPAddressClaims) error
+
+	// Teardown tries to cleanup orphaned IPAddressClaims by checking if the corresponding IPs are still in use in vSphere.
+	// It identifies IPAddressClaims via labels.
+	Teardown(ctx context.Context, folderName string, vSphereClient *govmomi.Client) error
+}
+
+// New returns an ipam.Helper. If e2eIPAMKubeconfig is an empty string or skipCleanup is true
+// it will return a noop helper which does nothing so we cann fallback on setting environment variables.
+func New(e2eIPAMKubeconfig string, labels map[string]string, skipCleanup bool) (Helper, error) {
+	if len(labels) == 0 {
+		return nil, fmt.Errorf("expecting labels to be set to prevent deletion of other IPAddressClaims")
+	}
+
+	if e2eIPAMKubeconfig == "" || skipCleanup {
```

Comment:
I would have assumed skipCleanup only skips the deletion of the IPs, but this also skips the creation?